doc/BMIconventions.md

```diff
@@ -0,0 +1,313 @@
+
+## Calling `set_value` on output variables
+
+The ngen process should not call `set_value` on an output variable. Attempting to create a formulation that connects forcing or another module's output to an input name that is in `get_output_var_names` risks undefined behavior.
+
+## Calling `get_value` on input variables
+
+The ngen process *may* call `get_value` on a variable that is returned from `get_input_var_names`. A model *should* return the last value provided to the input variable, unmodified, in this case. At a minimum, a model *must* not crash if this occurs, and should return some reasonable value, rather than throwing an exception or returning BMI_FAILURE (throwing an exception or returning BMI_FAILURE *is* however an appropriate and preferred response if no value has yet been set for the variable).
+
+## Input and output with the same name not permitted
+
+A single variable name cannot be published in both `get_var_input_names` and `get_var_output_names`--this will cause a validation error. Furthermore, it is not presently possible to work around this situation with variable name mapping, because <!-- TODO: This may change with a config file format refactor? --> a variable name mapping does not specify whether it is an input or output variable and could therefore not map only one or the other.
+
+# Unpublished variables
+
+As discussed above, it is not defined by the BMI specification that you may *only* call `get_value` and `set_value` on variable names returned by either `get_input_var_names` or `get_output_var_names`. We refer to variables that appear in neither list--but are understood by a BMI module--as "unpublished" variables. Because ngen uses the input and output variable listings to validate formulations, there are many use cases for such variables, for instance:
+
+* Initialization parameters
+* Calibratable parameters
+* State inspection (for serialization or debugging)
+* Metadata variables (spatial reference information, hyperparameters, etc.) 
+
+These and other uses may be applied by ngen. Specifically, ngen will make use of specified, reserved unpublished variables enumerated later in this document. Therefore, a module should be able to communicate via unpublished variables or at least not malfunction in the case of unpublished variables being used, as described below.
+
+Note that there is currently no discoverability mechanism for unpublished variables, they must be described in documentation (such as this document or a module's  documentation) and often manually implemented or configured for use (such as in a realization config file).
+
+## Unknown/unexpected variables passed to `set_value`
```

Comment:
Umm...I'm not sure I agree with this statement.  Would be happy to discuss in more detail.

Reply:
The header? Or the section comments? Sure, here, or hit me up on Slack.

Reply:
Why should `set_value` not return `BMI_FAILURE`?  If ngen is using unpublished conventional variables, these are essentially OPTIONAL and it *should* explicitly check to ensure that `BMI_SUCCESS` was returned from `set_value` to ensure that it can properly assume the semantics are valid.  `BMI_FAILURE` would indicate that those assumptions are not valid.  That is how calibration parameters are handled right now -- `BMI_FAILURE` just assumes that variable is not calibratable.  When we get to grid conventions, it is almost imperative for `BMI_FAILURE` to be returned if a call to `set_value( 'grid_X_shape')` fails...this allows the model engine to know that the grid cannot be manipulated by the framework, and it can test if the model implements the correct grid functions and their values align in a meaningful fashion.

Reply:
Ah. Okay, this part I was less sure about. Generally, I wasn't sure how ngen would handle `set_value` returning `BMI_FAILURE` and whether it would consider that an error for a convention variable. You are probably right here, especially if the ngen code already expects these semantics.

The only thing left in my mind is, if the module supports the variable, but encounters an unexpected problem, what then? For C++/Python we can recommend specific exceptions but we're kindof out of options for C/Fortran. Probably no good answer, but it's a likely scenario. Will it be important to distinguish between support+failure vs. nosupport?

Reply:
Nope, not without enumerated error codes...BMI 3.0 recommendation =)